neuraxle/steps/data.py

```diff
@@ -0,0 +1,162 @@
+"""
+Data Steps
+====================================
+You can find here steps that take action on data.
+
+..
+    Copyright 2019, Neuraxio Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+"""
+import random
+from typing import Iterable
+
+from neuraxle.base import BaseStep, MetaStepMixin, NonFittableMixin, ExecutionContext
+from neuraxle.data_container import DataContainer
+from neuraxle.steps.output_handlers import InputAndOutputTransformerMixin
+
+
+class DataShuffler(NonFittableMixin, InputAndOutputTransformerMixin, BaseStep):
+    """
+    Data Shuffling step that shuffles data inputs, and expected_outputs at the same time.
+
+    .. code-block:: python
+
+        p = Pipeline([
+            TrainOnlyWrapper(DataShuffler(seed=42, increment_seed_after_each_fit=True)),
+            TrainOnlyWrapper(EpochRepeater(ForecastingPipeline(), epochs=EPOCHS))
+            TestOnlyWrapper(ForecastingPipeline())
+        ])
+
+    .. warning::
+        You probably always want to wrap this step by a :class:`TrainOnlyWrapper`
+
+    .. seealso::
+        :class:`EpochRepeater`,
+        :class:`TrainOnlyWrapper`,
+        :class:`InputAndOutputTransformerMixin`,
+        :class:`BaseStep`
+    """
+
+    def __init__(self, seed, increment_seed_after_each_fit=True):
+        InputAndOutputTransformerMixin.__init__(self)
+        BaseStep.__init__(self)
+        self.seed = seed
+        self.increment_seed_after_each_fit = increment_seed_after_each_fit
+
+    def transform(self, data_inputs):
+        """
+        Shuffle data inputs, and expected outputs.
+
+        :param data_inputs: (data inputs, expected outputs) tuple to shuffle
+        :return:
+        """
+        if self.increment_seed_after_each_fit:
+            self.seed += 1
+
+        di, eo = data_inputs
+        data = list(zip(di, eo))
+        random.Random(self.seed).shuffle(data)
+
+        data_inputs_shuffled, expected_outputs_shuffled = list(zip(*data))
+
+        return list(data_inputs_shuffled), list(expected_outputs_shuffled)
+
+
+class EpochRepeater(MetaStepMixin, BaseStep):
+    """
+    Repeat wrapped step fit, or transform for the number of epochs passed in the constructor.
+
+    .. code-block:: python
+
+        p = Pipeline([
+            TrainOnlyWrapper(DataShuffler(seed=42, increment_seed_after_each_fit=True)),
+            TrainOnlyWrapper(EpochRepeater(ForecastingPipeline(), epochs=EPOCHS))
+            TestOnlyWrapper(ForecastingPipeline())
```

Comment:
Same comment here.

Reply:
the epoch repeater should only have a different behavior in test mode. removing the TrainOnlyWrapper, and TestOnlyWrapper. dead code